Tests and regex refactored

CFGrote opened the pull request:
This PR fixes various tests and refactors the regex used to check predicates in
the BaseLine model.

List of changes:
f21ff6209: The new regex to check for valid predicates is now:
```
predicate: str = Field(
        regex=r"^(P\d+)|([ADE]{1}[a-z]{2})$",
        description="Either a predicate, Label, Alt. Label, or Description",
    )
```
The old regex did not match labels other than in English language ("en"), the new regex matches all two-letter language codes. To be fair, I was too lazy to make a list of actually valid language codes so I only check for two lower case letters.

dfef1bbb9:
while testing locally I needed a setup.py. This commit adds setup.py to .gitignore.

08677bf29: fixes various issues detected by flake8

ca7cf6a72d: fixes deprecated syntax in doc/source/conf.py reported as error by docs-test

105507e76: pins sphinx version to <7.0 as later versions raise a `Unknown error "style is undefined"` in the docs-test step.

Commit:
Update model.py

- Edit regex for predicate, now also accepts alt. labels.

Commit:
Adding setup.py to gitignore.

Commit:
Fixing flake8 issues.

Commit:
Update sphinx conf.

- fix pre 1.0 syntax for intersphinx_mapping.

Commit:
Attempting to limit sphinx version to <7.

cthoyt commented:
Hi @CFGrote, thank you for the contribution! I will leave some specific code comments directly

cthoyt commented on src/quickstatements_client/model.py:
```diff
@@ -169,8 +169,8 @@ class BaseLine(BaseModel):
     predicate: str = Field(
-        regex=r"^(P\d+)|(Len)|(Den)$",
-        description="Either a predicate, `Len` (label), or `Den` (description)",
+        regex=r"^(P\d+)|([ADE]{1}[a-z]{2})$",
```
> The old regex did not match labels other than in English language ("en"), the new regex matches all two-letter language codes. To be fair, I was too lazy to make a list of actually valid language codes so I only check for two lower case letters.

I had a really hard time figuring out how quickstatements worked in the beginning and I want to make sure that the package explicitly explains how things work. Therefore, I would ask if you want to extend this, then you write everything out in full, with code comments. `[ADE]{1}[a-z]{2}` unfortunately doesn't mean anything, and the changes to the description string serve to obfuscate what's going on here, so I can't accept these changes as is

One way to do that would be to create a list or dictionary data structure with all of these parts in it, then assemble the regex programatically.

CFGrote replied:
by "dictionary data structure" you mean a dict of valid two-letter language codes?

cthoyt replied:
Ohhh I see what's going on here. The `en` at the end of `Len` is a two letter code. As far as I can remember, I did not see this anywhere, and it didn't even occur to me. So it can be `L{two letter code}` for label, `D{two letter code}` for description, and something else for Alts?  How did you get to ADE?

Maybe do something like this

```python
two_letter_codes = [
    "en",
    # fill in, or get programatically somehow
]

regex_part = "|".join(two_letter_codes)
regex = rf"^(P\d+)|([ADL]({regex_part}))$"
```

CFGrote replied:
Source:  https://www.wikidata.org/w/index.php?title=Help:QuickStatements&section=6#Adding_labels,_aliases,_descriptions_and_sitelinks

I'll work out something along your suggestion, thanks.

the "E" is a mistake, should be "L". That's fixed in a subsequent commit.

cthoyt commented on .gitignore:
```diff
@@ -899,3 +899,4 @@ FodyWeavers.xsd
 # End of https://www.toptal.com/developers/gitignore/api/macos,linux,windows,python,jupyternotebooks,jetbrains,pycharm,vim,emacs,visualstudiocode,visualstudio
 
 scratch/
+setup.py
```
why would you need a setup.py? This might mean you're using a very very old version of setuptools, so I would suggest upgrading this instead of changing the .gitignore. In fact, setup.py might not even work anymore in the near future

CFGrote replied:
true, my bad

Commit:
Revert "Adding setup.py to gitignore."

This reverts commit dfef1bbb951e9707c799482dfa579776c3f54792.

cthoyt commented:
@CFGrote please enable admins to push to your branch, this can be done by checking the box on the right side of the PR page

Commit:
Adding test for loading the ISO639 language codes.

Commit:
Adding module that holds the ISO639 language codes and import in model.py

Commit:
Moving file

Commit:
Implementing the language code regex and testing.

CFGrote committed:
Merge branch 'main' into main

Commit:
Fixing typo in regex (E should be L).

CFGrote commented:
> @CFGrote please enable admins to push to your branch, this can be done by checking the box on the right side of the PR page

sorry, I don't see where I would check that box??

cthoyt commented:
example:
![Uploading Screenshot 2023-06-30 at 00.03.41.png…]()

CFGrote commented:
No sorry, there's nothing like that on my site.

cthoyt committed:
Consolidate and streamline tests

cthoyt committed:
Improve language codes

1. Rename this from cryptic ISO639 to just "language_codes"
2. Don't hardcode data in Python file - use JSON instead. Later, we can encode the SPARQL query that generates this list for full transparency
3. Fully use the description field in the data model instead of floating code comments
4. Don't add a random variable "lang_patterns" in the class body
5. Simplify regex by not duplicating ADL for each language code
6. Improve description text

CFGrote committed:
Merge pull request #1 from cthoyt/cleanup-carsten

Cleanup pull request

CFGrote committed:
Update test_models.py

Fixing typo in language code

cthoyt commented:
@CFGrote thanks you for bearing with me and addressing all of the feedback. This is great! Liebe Grüße aus Bonn :)
